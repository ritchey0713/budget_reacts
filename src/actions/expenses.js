import uuid from "uuid"
import database from "../firebase/firebase"

// add expense 
export const addExpense = (expense) => {
  return {
    type: "ADD_EXPENSE",
    expense  
  }
}

export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = "", 
      note = "", 
      amount = 0, 
      createdAt = 0
    } = expenseData
    const expense = {description, note, amount, createdAt}
    database.ref("expenses").push(expense)
      .then((ref) => {
        dispatch(addExpense({
          id: ref.key,
          ...expense
        }))
      })
  }
}

export const removeExpense = ({ id } = {}) => {
  return {
    type: "REMOVE_EXPENSE",
    id
  }
}

export const editExpense = (id, updates) => {
  return {
    type: "EDIT_EXPENSE",
    id, 
    updates
  }
}


//SET expenses
export const setExpenses = (expenses) => {
  return {
    type: 'SET_EXPENSES',
    expenses
  }
};

export const startSetExpenses = () => {
 return (dispatch) => {
   return database.ref("expenses").once("value")
    .then((snapShot) => {
      const expenses = []
      snapShot.forEach((childSnapShot) => {
        expenses.push({
          id: childSnapShot.key,
          ...childSnapShot.val()
        })
      })
      dispatch(setExpenses(expenses))
    })
 }
};

// export const startSetExpenses = () => {
//   return (dispatch) => {
//     return database.ref("expenses").on("value", (snapShot) => {
//       const expenses = []
//       snapShot.forEach((childSnapShot) => {
//         return expenses.push({
//           id: childSnapShot.key,
//           ...childSnapShot.val()
//         })
//       })
//       dispatch(setExpenses(expenses))
//     })
//   }
// }
