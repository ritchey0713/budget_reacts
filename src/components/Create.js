import React from "react" 
import ExpenseForm from "./ExpenseForm"
import { connect } from 'react-redux'
import { addExpense } from "../actions/expenses"

const CreateExpense = (props) => {
  return (
    <div>
      <h1>Add new budget</h1>
      <ExpenseForm
        reduxOnSubmit={(expense) => {
          console.log(expense)
          props.dispatch(addExpense(expense))
          props.history.push("/")
        }}
      />
    </div>
  )
}

export default connect()(CreateExpense)