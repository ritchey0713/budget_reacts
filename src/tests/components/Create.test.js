import React from "react"
import { shallow } from "enzyme"
import { CreateExpense } from "../../components/Create.js"
import expenses from "../fixtures/expenses"

test("should render create expense page", () => {
  const onSubmit = jest.fn()
  const history = {push: jest.fn()}
  const wrapper = shallow(<CreateExpense startAddExpense={onSubmit} history={history}/>)
  expect(wrapper).toMatchSnapshot()
})

test("should handle onSubmit", () => {
  const onSubmit = jest.fn()
  const history = {push: jest.fn()}
  const wrapper = shallow(<CreateExpense startAddExpense={onSubmit} history={history}/>)
  wrapper.find("ExpenseForm").prop("reduxOnSubmit")(expenses[0])

  expect(history.push).toHaveBeenLastCalledWith("/")
  expect(onSubmit).toHaveBeenLastCalledWith(expenses[0])
})