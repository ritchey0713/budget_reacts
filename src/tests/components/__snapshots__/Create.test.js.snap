// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render create expense page 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Add new budget
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      reduxOnSubmit={[Function]}
    />
  </div>
</div>
`;
